recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -312,15 +343,79 @@ class ManagedFirmwareUpdate(FirmwareUpdate):
             while choice > len(firmware_names) or choice < 1:
                 print("Out of range, please reinput your choice:")
                 choice = int(input("-> "))
-            
+
             firmware_name = firmware_names[choice - 1]
 
         print("Will write {} to the SPI flash".format(firmware_name))
 
         self.firmware.seek(0)
-        self.firmware = tarfile.open(fileobj=self.firmware).extractfile(firmware_name)
-        super().update_firmware()
+        with tarfile.open(fileobj=self.firmware) as f:
+            self.firmware = f.extractfile(firmware_name)
+            f.extract(member=self.uboot_default_env_file_name, path=self.extract_path)
+            super().update_firmware()
 
+    def update_uboot_env(self, env_list):
+        '''
+        restore uboot env
+        retrive the uboot default env file belong to the upgrade firmware.
+        makeup the restore env binary and flash to the env mtd partition
+        '''
+        import shutil
+        # /dev/mtd3     /dev/mtd4
+        env_partition_list = [3, 4]
+        uboot_default_env_file = os.path.join(self.extract_path, self.uboot_default_env_file_name)
+        uboot_env_assemble_file =  os.path.join(self.extract_path, "env_assemble_file")
+        assert os.path.isfile(uboot_default_env_file)
+        # assemble the env
+        shutil.copy(uboot_default_env_file, uboot_env_assemble_file)
+        with open(uboot_env_assemble_file, encoding="utf-8", mode="a") as file:
+            for value in env_list:
+                file.write(value)
+                file.write("\n")
+
+        with tempfile.NamedTemporaryFile(mode="rb") as env_default_binary:
```

Comment:
Why `mode="rb"`? shouldn't be `mode="wb"`?

Reply:
remove the mode "rb" use the default `w+b` mode